src/attach-shadow.js

```diff
@@ -239,25 +298,140 @@ ShadyRoot.prototype._updateChildNodes = function(container, children) {
-  return this._distributor.insertionPointTag;
+ShadyRoot.prototype._addSlots = function(slots) {
+  let slotNamesToSort;
+  for (let i=0; i < slots.length; i++) {
+    let slot = slots[i];
+    // ensure insertionPoints's and their parents have logical dom info.
+    // save logical tree info
+    // a. for shadyRoot
+    // b. for insertion points (fallback)
+    // c. for parents of insertion points
+    slot.__shady = slot.__shady || {};
+    recordChildNodes(slot);
+    recordChildNodes(slot.parentNode);
+    let name = this._nameForSlot(slot);
+    if (this._slotMap && this._slotMap[name]) {
+      slotNamesToSort = slotNamesToSort || {};
+      slotNamesToSort[name] = true;
+    }
+    this._slotList.push(slot);
+  }
+  // TODO(sorvell): add tests to verify slots add/removing slots
+    // with the same name (or catchall)
+  if (slotNamesToSort) {
+    for (let n in slotNamesToSort) {
+      let slotsForName = this._extractSlotsOfName(n);
+      let sorted = this._sortSlots(slotsForName);
+      this._slotList.push(...sorted);
+    }
+  }
+  this._updateSlotMap();
 }
 
-ShadyRoot.prototype.hasInsertionPoint = function() {
-  return Boolean(this._insertionPoints && this._insertionPoints.length);
+ShadyRoot.prototype._nameForSlot = function(slot) {
+  return slot['name'] || slot.getAttribute('name') || CATCHALL_NAME;
 }
 
-ShadyRoot.prototype._getInsertionPoints = function() {
-  if (!this._insertionPoints) {
-    this.updateInsertionPoints();
+ShadyRoot.prototype._extractSlotsOfName = function(name) {
+  let slots;
+  for (let i=0; i < this._slotList.length; i++) {
+    let slot = this._slotList[i];
+    let n = this._nameForSlot(slot);
+    if (n == name) {
+      this._slotList.splice(i, 1);
+      i--;
+      slots = slots || [];
+      slots.push(slot);
+    }
+  }
+  return slots;
+}
+
+ShadyRoot.prototype._sortSlots = function(slots) {
+  return slots.sort((a, b) => {
+    let listA = ancestorList(a);
+    let listB = ancestorList(b);
+    for (var i=0; i < listA.length; i++) {
+      let nA = listA[i];
+      let nB = listB[i];
+      if (nA !== nB) {
+        let c$ = Array.from(nA.parentNode.childNodes);
+        return c$.indexOf(nA) - c$.indexOf(nB);
+      }
+    }
+  });
+}
+
+function ancestorList(node) {
+  let ancestors = [];
+  do {
+    ancestors.unshift(node);
+  } while ((node = node.parentNode));
+  return ancestors;
+}
+
+function contains(container, node) {
+  while (node) {
+    if (node == container) {
+      return true;
+    }
+    node = node.parentNode;
+  }
+}
+
+ShadyRoot.prototype._removeContainerSlots = function(container) {
+  let didRemove;
+  for (let i=0; i<this._slotList.length; i++) {
+    let slot = this._slotList[i];
+    if (contains(container, slot)) {
+      this._slotList.splice(i, 1);
+      this._removeAssignedNodes(slot);
+      i--;
+      didRemove = true;
+    }
+  }
+  this._updateSlotMap();
+  return didRemove;
+}
+
+ShadyRoot.prototype._removeSlot = function(slot) {
```

Comment:
This is only used by `distributeAttributeChange` in logical-mutation, when that function could just use `root._updateSlotMap()`. Consider deleting.